flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/SortOperationFactory.java

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.expressions.ApiExpressionDefaultVisitor;
+import org.apache.flink.table.expressions.CallExpression;
+import org.apache.flink.table.expressions.Expression;
+import org.apache.flink.table.expressions.ExpressionBridge;
+import org.apache.flink.table.expressions.PlannerExpression;
+import org.apache.flink.table.plan.logical.Limit;
+import org.apache.flink.table.plan.logical.LogicalNode;
+import org.apache.flink.table.plan.logical.Sort;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.singletonList;
+import static org.apache.flink.table.expressions.BuiltInFunctionDefinitions.ORDERING;
+import static org.apache.flink.table.expressions.BuiltInFunctionDefinitions.ORDER_ASC;
+
+/**
+ * Utility class for creating a valid {@link Sort} operation.
+ */
+@Internal
+public class SortOperationFactory {
+
+	private final boolean isStreaming;
+	private final ExpressionBridge<PlannerExpression> expressionBridge;
+	private final OrderWrapper orderWrapper = new OrderWrapper();
+
+	public SortOperationFactory(
+			ExpressionBridge<PlannerExpression> expressionBridge,
+			boolean isStreaming) {
+		this.expressionBridge = expressionBridge;
+		this.isStreaming = isStreaming;
+	}
+
+	/**
+	 * Creates a valid {@link Sort} operation.
+	 *
+	 * <p><b>NOTE:</b> if the collation is not explicitly specified for any expression, it is wrapped in a
+	 * default ascending order
+	 *
+	 * @param orders expressions describing order,
+	 * @param child relational expression on top of which to apply the sort operation
+	 * @return valid sort operation
+	 */
+	public Sort createSort(List<Expression> orders, TableOperation child) {
+		failIfStreaming();
+
+		List<PlannerExpression> convertedOrders = orders.stream()
+			.map(f -> f.accept(orderWrapper))
+			.map(expressionBridge::bridge)
+			.collect(Collectors.toList());
+		return new Sort(convertedOrders, (LogicalNode) child);
+	}
+
+	/**
+	 * Creates a valid {@link Limit} operation.
+	 *
+	 * @param offset offset to start from
+	 * @param fetch number of records to fetch
+	 * @return valid limit operation
+	 */
+	public Limit createLimit(int offset, int fetch, TableOperation child) {
+		failIfStreaming();
+
+		if (!(child instanceof Sort)) {
+			throw new ValidationException("Limit operator must be preceded by an OrderBy operator.");
+		}
+		if (offset < 0) {
+			throw new ValidationException("Offset should be greater than or equal to zero.");
+		}
+
+		return new Limit(offset, fetch, (LogicalNode) child);
+	}
+
+	private void failIfStreaming() {
+		if (isStreaming) {
+			throw new ValidationException("Limit on stream tables is currently not supported.");
```

Comment:
nit: rename exception to "A limit operation on unbounded tables is currently not supported."

we should be consistent with the concepts.